### PR TITLE
Node Identity

### PR DESCRIPTION
**Pull request by bryanchriswhite**
This change is [Reviewable](https://reviewable.io/reviews/storj/storj/193)

### CODE REVIEW

File: cmd/hc/main.go [hunk 1]
```diff
@@ -41,7 +41,10 @@ var (
 		Overlay   overlay.Config
 	}
 	setupCfg struct {
-		BasePath string `default:"$CONFDIR" help:"base path for setup"`
+		BasePath    string `default:"$CONFDIR" help:"base path for setup"`
+		Concurrency uint   `default:"4" help:"number of concurrent workers for certificate authority generation"`
+		CA          provider.CAConfig
+		Identity    provider.IdentityConfig
```
Comment (jtolds) [inline cmd/hc/main.go]: why are these last two here? CAConfig defines paths, a difficulty, and a version, but all of those should be calculable from BasePath, Concurrency, and setup should just choose the latest version. the idea behind setupCfg is everything you need to generate a good config.

File: cmd/hc/main.go [hunk 2]
```diff
@@ -65,8 +68,19 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
 		return err
 	}
 
-	identityPath := filepath.Join(setupCfg.BasePath, "identity")
-	_, err = peertls.NewTLSFileOptions(identityPath, identityPath, true, false)
+	// GenerateCA CA
+	ca := provider.GenerateCA(context.Background(), setupCfg.CA.Difficulty, 4)
```
Comment (jtolds) [inline cmd/hc/main.go]: should use process.Ctx(cmd) for context here. you should never need to use context.Background going forward except for in very exceptional cases

Comment (jtolds) [reply cmd/hc/main.go]: also, this can't return an error?

File: cmd/hc/main.go [hunk 3]
```diff
@@ -65,8 +68,19 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+	fi, err := ca.GenerateIdentity()
+	if err != nil {
+		return err
+	}
+
+	err = setupCfg.CA.Save(ca)
```
Comment (jtolds) [inline cmd/hc/main.go]: this is surprising to me. what was basepath for?

should ca have a save method that takes a path?

File: cmd/hc/main.go [hunk 4]
```diff
@@ -65,8 +68,19 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
+	if err != nil {
+		return err
+	}
+
+	err = setupCfg.Identity.Save(fi)
```
Comment (jtolds) [inline cmd/hc/main.go]: same

File: pkg/peertls/peertls.go [hunk 5]
```diff
@@ -34,131 +41,137 @@ var (
...
+		signer = &tls.Certificate{
+			PrivateKey: k,
+			Leaf: &x509.Certificate{
+				PublicKey: &k.PublicKey,
+			},
+		}
+	}
+	sk, ok := signer.Leaf.PublicKey.(*ecdsa.PublicKey)
+	if !ok {
+		return nil, ErrUnsupportedKey.New("%s", reflect.TypeOf(sk))
```
Comment (jtolds) [inline pkg/peertls/peertls.go]: you can just do `ErrUnsupportedKey.New("%T", sk)` for the same effect

File: pkg/peertls/peertls.go [hunk 6]
```diff
@@ -34,131 +41,137 @@ var (
...
+		template,
+		parentTemplate,
+		parent.Certificate,
+		sk,
+		k,
+		signer.PrivateKey,
+	)
+	if err != nil {
+		return nil, ErrGenerate.Wrap(err)
+	}
```
Comment (jtolds) [inline pkg/peertls/peertls.go]: could just do `return createCert(...)` and have createCert return an ErrGenerate wrapped error

File: pkg/peertls/peertls.go [hunk 7]
```diff
@@ -34,131 +41,137 @@ var (
...
+		c, err := parseCertificateChains(chain)
+		if err != nil {
+			return err
 		}
 
-		if !isValid {
-			return ErrVerifyPeerCert.New("certificate chain signature verification failed")
+		for _, n := range next {
+			if n != nil {
+				if err := n(nil, [][]*x509.Certificate{c}); err != nil {
```
Comment (jtolds) [inline pkg/peertls/peertls.go]: huh, only one cert, no bytes?

File: pkg/peertls/peertls_test.go [hunk 8]
```diff
@@ -5,475 +5,94 @@ package peertls
...
+		ca, err := x509.ParseCertificate(l.Certificate[1])
+		if err != nil {
+			return err
 		}
+		lf, err := x509.ParseCertificate(l.Certificate[0])
+		if err != nil {
+			return err
+		}
+
+		switch true {
```
Comment (jtolds) [inline pkg/peertls/peertls_test.go]: can't you just do `switch {`?

File: pkg/provider/identity.go [hunk 9]
```diff
@@ -6,27 +6,46 @@ package provider
...
+	Cert *x509.Certificate
+	// The ID is calculated from the CA cert.
+	ID nodeID
+}
+
 // PeerIdentity represents another peer on the network.
 type PeerIdentity struct {
 	// CA represents the peer's self-signed CA. The ID is taken from this cert.
-	CA *x509.Certificate
+	CA CertificateAuthority
```
Comment (jtolds) [inline pkg/provider/identity.go]: hmm, but a CertificateAuthority has a PrivateKey field. you definitely won't have that for a peer

File: pkg/provider/identity.go [hunk 10]
```diff
@@ -35,26 +54,235 @@ type PeerIdentity struct {
 type FullIdentity struct {
 	PeerIdentity
 	PrivateKey crypto.PrivateKey
+}
 
-	todoCert *tls.Certificate // TODO(jt): get rid of this and only use the above
+type CAConfig struct {
+	CertPath   string `help:"path to the certificate chain for this identity" default:"$CONFDIR/identity.leaf.cert"`
+	KeyPath    string `help:"path to the private key for this identity" default:"$CONFDIR/identity.leaf.key"`
+	Difficulty uint16 `default:"24" help:"minimum difficulty for identity generation"`
```
Comment (jtolds) [inline pkg/provider/identity.go]: once a CA is created, you don't need the difficulty anymore? difficulty seems like a creation thing, not a property of an existing CA

File: pkg/provider/identity.go [hunk 11]
```diff
@@ -35,26 +54,235 @@ type PeerIdentity struct {
...
+	Version    string `help:"semantic version of CA storage format"`
```
Comment (jtolds) [inline pkg/provider/identity.go]: version seems like it's related to loading, but not to the in memory representation, right? after you load a CA the version doesn't matter anymore?

File: pkg/provider/identity.go [hunk 12]
```diff
@@ -35,26 +54,235 @@ type PeerIdentity struct {
...
 }
 
 // IdentityConfig allows you to run a set of Responsibilities with the given
 // identity. You can also just load an Identity from disk.
 type IdentityConfig struct {
 	CertPath string `help:"path to the certificate chain for this identity" default:"$CONFDIR/identity.leaf.cert"`
 	KeyPath  string `help:"path to the private key for this identity" default:"$CONFDIR/identity.leaf.key"`
+	Version  string `help:"semantic version of identity storage format"`
```
Comment (jtolds) [inline pkg/provider/identity.go]: same. when you load a key, you'll figure this out. why store it in the struct? it doesn't matter after loading.

File: pkg/provider/identity.go [hunk 13]
```diff
@@ -35,26 +54,235 @@ type PeerIdentity struct {
...
 }
 
-// LoadIdentity loads a FullIdentity from the given configuration
-func (ic IdentityConfig) LoadIdentity() (*FullIdentity, error) {
-	pi, err := FullIdentityFromFiles(ic.CertPath, ic.KeyPath)
+// GenerateCA creates a new full identity with the given difficulty
+func GenerateCA(ctx context.Context, difficulty uint16, concurrency uint) *CertificateAuthority {
+	if ctx == nil {
+		ctx = context.Background()
+	}
```
Comment (jtolds) [inline pkg/provider/identity.go]: the go context style says you should never ever ever pass a context as nil, and thus you should never have to worry about checking if it's nil

File: pkg/provider/identity.go [hunk 14]
```diff
@@ -35,26 +54,235 @@ type PeerIdentity struct {
...
+	ctx, cancel := context.WithCancel(ctx)
+
+	caC := make(chan CertificateAuthority, 1)
+	for i := 0; i < int(concurrency); i++ {
+		go generateCAWorker(ctx, difficulty, caC)
```
Comment (jtolds) [inline pkg/provider/identity.go]: is there no way any of these can fail in a bad way?

File: pkg/provider/identity.go [hunk 15]
```diff
@@ -35,26 +54,235 @@ type PeerIdentity struct {
...
+		CA: CertificateAuthority{
+			Cert: ca,
+			ID:   i,
+		},
+		Leaf: leaf,
+	}, nil
+}
+
+// VerifyPeerIdentityFunc returns a function to use with `tls.Certificate.VerifyPeerCertificate`
+// that verifies the peer identity satisfies the minimum difficulty
```
Comment (jtolds) [inline pkg/provider/identity.go]: just want to reiterate that regardless of the minimum difficulty of a node you will want to store to, there will always be cases where you'll want to allow ssl communication with nodes that have difficulty 0, so in practice we can't ever use this.

File: pkg/provider/utils.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
+package provider
```
Comment (jtolds) [inline pkg/provider/utils.go]: missing copyright

File: pkg/provider/utils.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+				continue
+			}
+
+			c, err := peertls.Generate(ct, nil, nil, nil)
+			if err != nil {
+				zap.S().Error(err)
+				continue
+			}
+
+			i, err := idFromCert(c.Leaf)
```
Comment (jtolds) [inline pkg/provider/utils.go]: you probably can avoid generating a cert at all - all you need is a public key, right? so the tightest loop is going to be generating public keys until you find one, then constructing a cert around it

**Comment by jtolds** [comment]
think we can find a way to break this pr into even smaller prs? merge one feature first, make sure it's settled, merge another piece that depends on it, etc?

File: cmd/hc/main.go (hunk 1 above)
Comment (bryanchriswhite) [reply cmd/hc/main.go]: What if you want to specify separate paths for the key vs cert for example?

File: pkg/peertls/peertls.go (hunk 5 above)
Comment (bryanchriswhite) [reply pkg/peertls/peertls.go]: thanks for the tip!

File: pkg/provider/identity.go (hunk 13 above)
Comment (bryanchriswhite) [reply pkg/provider/identity.go]: noted - i'll clean it up

File: pkg/provider/identity.go (hunk 15 above)
Comment (bryanchriswhite) [reply pkg/provider/identity.go]: What about just passing a 0 difficulty?: https://sourcegraph.com/github.com/storj/storj@f5f7f15e9a2eba1939bab25be74cebc282ead81e/-/blob/pkg/provider/provider.go#L34

Comment (jtolds) [reply pkg/provider/identity.go]: why have it though? it will always be a zero difficulty

File: pkg/provider/identity.go (hunk 14 above)
Comment (bryanchriswhite) [reply pkg/provider/identity.go]: yea but I could use some input on how to handle it best...

File: cmd/captplanet/setup.go [hunk 18]
```diff
@@ -54,8 +58,15 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
 	if err != nil {
 		return err
 	}
-	identPath := filepath.Join(hcPath, "ident")
-	_, err = peertls.NewTLSFileOptions(identPath, identPath, true, true)
+	// TODO: handle setting base path *and* identity file paths via args
+	// NB: if base path is set this overrides identity and CA path options
+	if setupCfg.BasePath != defaultConfDir {
```
Comment (jtolds) [inline cmd/captplanet/setup.go]: if setupCfg.BasePath == defaultConf, setupCfg.HCCA.CertPath and setupCfg.GWCA.CertPath are the same thing, which is bad. i think for setup it really will be simpler to just not worry about letting people configure these separately

File: cmd/captplanet/setup.go [hunk 19]
```diff
@@ -66,8 +77,21 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
 		if err != nil {
 			return err
 		}
-		identPath = filepath.Join(farmerPath, "ident")
-		_, err = peertls.NewTLSFileOptions(identPath, identPath, true, true)
+		farmerCA := provider.CASetupConfig{
+			CertPath: filepath.Join(farmerPath, "ca.cert"),
+			KeyPath:  filepath.Join(farmerPath, "ca.key"),
+		}
```
Comment (jtolds) [inline cmd/captplanet/setup.go]: looks like there isn't a way to specify difficulty and a bunch of other things.

here's my recommendation: just manually make these fields part of setupCfg directly:

```
	Difficulty  uint64 `help:"minimum difficulty for identity generation" default:"24"`
	Timeout     string `help:"timeout for CA generation; golang duration string (0 no timeout)" default:"5m"`
	Overwrite   bool   `help:"if true, existing CA certs AND keys will overwritten" default:"false"`
	Concurrency uint   `help:"number of concurrent workers for certificate authority generation" default:"4"`
```

then manually fill in all the setupconfigs using those values and basepath.

File: cmd/captplanet/setup.go (hunk 18 above)
Comment (jtolds) [reply cmd/captplanet/setup.go]: see comment below about farmer ca setup

File: cmd/captplanet/setup.go (hunk 19 above)
Comment (jtolds) [reply cmd/captplanet/setup.go]: or alternatively, make the paths not be part of the casetup struct, and instead be an argument

File: cmd/captplanet/setup.go [hunk 20]
```diff
@@ -78,19 +76,21 @@ func cmdSetup(cmd *cobra.Command, args []string) (err error) {
...
 		}
 		farmerCA := provider.CASetupConfig{
-			CertPath: filepath.Join(farmerPath, "ca.cert"),
-			KeyPath:  filepath.Join(farmerPath, "ca.key"),
+			CertPath:    filepath.Join(farmerPath, "ca.cert"),
+			KeyPath:     filepath.Join(farmerPath, "ca.key"),
+			Difficulty:  setupCfg.FarmerCA.Difficulty,
+			Timeout:     setupCfg.FarmerCA.Timeout,
+			Overwrite:   setupCfg.FarmerCA.Overwrite,
+			Concurrency: setupCfg.FarmerCA.Concurrency,
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply cmd/captplanet/setup.go]: ha - good point

File: cmd/gw/main.go [hunk 21]
```diff
@@ -34,8 +34,11 @@ var (
 
 	runCfg   miniogw.Config
 	setupCfg struct {
-		BasePath  string `default:"$CONFDIR" help:"base path for setup"`
-		Overwrite bool   `default:"false" help:"whether to overwrite pre-existing configuration files"`
+		CA          provider.CASetupConfig
+		Identity    provider.IdentitySetupConfig
+		BasePath    string `default:"$CONFDIR" help:"base path for setup"`
+		Concurrency uint   `default:"4" help:"number of concurrent workers for certificate authority generation"`
```
Comment (jtolds) [inline cmd/gw/main.go]: aren't these in CASetupConfig now?

**Pull request merged**